### PR TITLE
[8.x] Add doesntHave method on the Collection & Lazy Collection classes

### PR DESCRIPTION
**Pull request by andresayej**
This PR adds `doesntHave()` method on the Collection & Lazy Collection classes.
The method is an inverse of the `has()` method.

The method aims to bring more human-like expression when checking for a key absence instead of doing it the typical PHP way: 
```php
<?php
$collection = collect(['id' => 1, 'first' => 'Hello', 'second' => 'World']);

if (! $collection->has('third')) {
...
```
With this PR you can do this:
```php
<?php
$collection = collect(['id' => 1, 'first' => 'Hello', 'second' => 'World']);

if ($collection->doesntHave('third')) {
...
```

The tests were run with the ` bin/test.sh` command contributed by @GrahamCampbell.
I've tried to follow the contribution guides as best as I can, but if I'm missing something do let me know.

As requested by @taylorotwell: https://github.com/laravel/framework/pull/32866#issuecomment-630253791

### CODE REVIEW

**Comment by netpok** [comment]
Since the target is master you should add it to the Enumerable interface too.

**Comment by andresayej** [comment]
> Since the target is master you should add it to the Enumerable interface too.

Thanks for the help @netpok, I think I've added it.
Since I'm making this for the first time, can you check if I've added it properly to the right class?

**Comment by netpok** [comment]
It looks okay to me.

File: tests/Support/SupportCollectionTest.php [hunk 1]
```diff
@@ -1678,6 +1678,18 @@ public function testHas($collection)
...
+    /**
+     * @dataProvider collectionClassProvider
+     */
+    public function testDoesntHave($collection)
+    {
+        $data = new $collection(['id' => 1, 'first' => 'Hello', 'second' => 'World']);
+        $this->assertFalse($data->doesntHave('first'));
+        $this->assertTrue($data->doesntHave('third'));
+        $this->assertFalse($data->doesntHave(['first', 'second']));
+        $this->assertTrue($data->doesntHave(['third', 'first']));
```
Comment (JosephSilber) [inline tests/Support/SupportCollectionTest.php:1690]: The `has` method returns `true` only if _all_ keys passed to it are present.

Shouldn't the `doesntHave` method also only return `true` if _all_ keys are missing?

File: tests/Support/SupportCollectionTest.php [hunk 2]
```diff
@@ -4154,6 +4166,18 @@ public function testHasReturnsValidResults($collection)
...
+     * @dataProvider collectionClassProvider
+     */
+    public function testDoesntHaveReturnsValidResults($collection)
+    {
+        $data = new $collection(['foo' => 'one', 'bar' => 'two', 1 => 'three']);
+        $this->assertFalse($data->doesntHave('foo'));
+        $this->assertFalse($data->doesntHave('foo', 'bar', 1));
+        $this->assertTrue($data->doesntHave('foo', 'bar', 1, 'baz'));
+        $this->assertTrue($data->doesntHave('baz'));
+    }
```
[the remark this answers is not in the document]
Comment (andresayej) [reply tests/Support/SupportCollectionTest.php:4179]: I was wondering about that too. However, I've done it because I saw that the `has` method has the same double test here: [Link](https://github.com/laravel/framework/blob/master/tests/Support/SupportCollectionTest.php#L4160).

[the remark this answers is not in the document]
Comment (andresayej) [reply tests/Support/SupportCollectionTest.php:4179]: Furthermore the `testHasReturnsValidResults()` test was added on 05.08.2019 which is newer than the `testHas()` test, so I thought that the test is important and maybe covers some newer edge case.

File: tests/Support/SupportCollectionTest.php (hunk 1 above)
Comment (andresayej) [reply tests/Support/SupportCollectionTest.php:1690]: I'm not sure how to achieve the described behavior, can I ask for help?

Comment (andresayej) [reply tests/Support/SupportCollectionTest.php:1690]: I've tried to do the following which should give the expected behavior but with no luck:
```php
<?php
    /**
     * Determine if an item does not exist in the collection by key.
     *
     * @param  mixed  $key
     * @return bool
     */
    public function doesntHave($key)
    {
        return ! $this->has($key);
    }
```

Comment (netpok) [reply tests/Support/SupportCollectionTest.php:1690]: I was thinking about this too, actually a better name for this should be something like hasMissing or similar.

This implementation checks till the first occurrence.

to check for all missing elements the condition of if needs to be negated, not the return value.
--
An other possible way would be to have a doesntHaveAny method implementing that feature.

I'm not sure about the correct terminology since English is not my native language:
- has: has all passed key
- doesntHave: has none of the keys
- hasAny: has one of the passed keys
- doesntHaveOne: missing at least one <-- Implemented by this PR

Comment (andresayej) [reply tests/Support/SupportCollectionTest.php:1690]: Actually if this is used:
```php
<?php
    /**
     * Determine if an item does not exist in the collection by key.
     *
     * @param  mixed  $key
     * @return bool
     */
    public function doesntHave($key)
    {
        return ! $this->has($key);
    }
```
it works as expected a true inverse of the `has` method. Also I've updated the test with the following which should prove it:
```php
<?php
    /**
     * @dataProvider collectionClassProvider
     */
    public function testDoesntHaveReturnsValidResults($collection)
    {
        $data = new $collection(['foo' => 'one', 'bar' => 'two', 1 => 'three']);
        $this->assertFalse($data->doesntHave('foo'));
        $this->assertFalse($data->doesntHave('foo', 'bar', 1));
        $this->assertFalse($data->doesntHave('foo', 'bar', 1, 'baz'));
        $this->assertTrue($data->doesntHave('baz'));
        $this->assertTrue($data->doesntHave(['baz', '2', 3]));
    }
```

Comment (andresayej) [reply tests/Support/SupportCollectionTest.php:1690]: @JosephSilber, @taylorotwell what do you think, should I make a new PR?

Comment (andresayej) [reply tests/Support/SupportCollectionTest.php:1690]: ping @JosephSilber @taylorotwell

**Comment by driesvints** [comment]
We've reverted this pr for the time being.
